src/bank.rs

```diff
@@ -146,147 +134,536 @@ pub struct LastIdEntry {
+        BankState {
+            checkpoints,
+            ..Bank::default()
+        }
     }
 
-    fn purge(&mut self, depth: usize) {
-        fn merge(into: &mut HashMap<Pubkey, Account>, purge: &mut HashMap<Pubkey, Account>) {
-            purge.retain(|pubkey, _| !into.contains_key(pubkey));
-            into.extend(purge.drain());
-            into.retain(|_, account| account.tokens != 0);
-        }
+    fn fork(&mut self) -> BankState {
+        let new = BankState::default();
 
-        while self.depth() > depth {
-            let (mut purge, _) = self.checkpoints.pop_back().unwrap();
+        self.num_children += 1;
 
-            if let Some((into, _)) = self.checkpoints.back_mut() {
-                merge(into, &mut purge);
-                continue;
-            }
-            merge(&mut self.accounts, &mut purge);
+        BankState {
+            accounts: HashMap::new(),
+            num_children: 0,
+            parent: self.last_entry_id,
+            last_tick_id_entries: self.last_tick_id_entries.clone(),
+            ..self // transaction_count, tick_height, last_tick_id
         }
     }
-    fn depth(&self) -> usize {
-        self.checkpoints.len()
-    }
+
+    //    fn rollback(&mut self) {
+    //        let (accounts, transaction_count) = self.checkpoints.pop_front().unwrap();
+    //        self.accounts = accounts;
+    //        self.transaction_count = transaction_count;
+    //    }
+    //
+    //    fn purge(&mut self, depth: usize) {
+    //        // accounts merge
+    //        fn merge(into: &mut HashMap<Pubkey, Account>, purge: &mut HashMap<Pubkey, Account>) {
+    //            purge.retain(|pubkey, _| !into.contains_key(pubkey));
+    //            into.extend(purge.drain());
+    //            into.retain(|_, account| account.tokens != 0);
+    //        }
+    //
+    //        while self.depth() > depth {
+    //            let (mut purge, _) = self.checkpoints.pop_back().unwrap();
+    //
+    //            if let Some((into, _)) = self.checkpoints.back_mut() {
+    //                merge(into, &mut purge);
+    //                continue;
+    //            }
+    //            merge(&mut self.accounts, &mut purge);
+    //        }
+    //    }
+    //
+    //    fn depth(&self, checkpoints: &HashMap<Hash, BankCheckpoint>) -> usize {
+    //        if let Some(parent) = self.parent {
+    //            checkpoints.get(parent).depth() + 1
+    //        } else {
+    //            0
+    //        }
+    //    }
 }
 
 /// Manager for the state of all accounts and contracts after processing its entries.
 pub struct Bank {
-    /// A map of account public keys to the balance in that account.
-    pub accounts: RwLock<Accounts>,
+    /// current bank state and all checkpoints
+    pub heads: RwLock<VecDeque<RwLock<BankState>>>,
 
-    /// FIFO queue of `last_id` items
-    last_ids: RwLock<LastIds>,
+    checkpoints: RwLock<HashMap<Hash, BankState>>,
```

Comment:
this is the interesting part, but has turned into a huge mess, where everything's now implemented by BankState...